src/System.Text.Primitives/System/Text/FormattingData.cs

```diff
@@ -12,19 +12,309 @@ public struct FormattingData
+                        return;
+                    }
+                }
+                else if (oldByteValue == null || oldByteValue < newByteValue)   // if there isn't a collision, we select the order we want to insert these nodes into
+                {                                                               // the list. This makes it so that our indices are sorted, which allows us to binary
+                    _childTrieNodes.Add(new TrieNode(oldByteValue, oldIndex));  // search the indices to recover info in O(logN).
+                    _childTrieNodes.Add(new TrieNode(newByteValue, newIndex));
+                }
+                else
+                {
+                    _childTrieNodes.Add(new TrieNode(newByteValue, newIndex));
+                    _childTrieNodes.Add(new TrieNode(oldByteValue, oldIndex));
+                }
+                node.ChildIndices.Clear();
+                node.ChildIndices.AddRange(new int[] { _childTrieNodes.Count - 2, _childTrieNodes.Count - 1 }); // we now have two branches coming from this node
+                return;
+            }
+            else    // if this is not a leaf, we need to add a new branch to the appropriate spot
+            {
+                var search = BinarySearch(node, level, newByteValue);   // we search to see if this value already exists in the branches
+
+                if (search.Item1)   // if it does exist within the branches
+                {
+                    CreateChildNode(_childTrieNodes[node.ChildIndices[search.Item2]], newIndex, level + 1); // we need to recurse down that branch
+                }
+                else    // if it doesn't exist, then we get to add a new branch to this node
+                {
+                    _childTrieNodes.Add(new TrieNode(newByteValue, newIndex));          // we create the new node
+                    node.ChildIndices.Insert(search.Item2, _childTrieNodes.Count - 1);  // and then insert at the appropriate index to keep the indices sorted
+                    return;
+                }
+            }
+            
+        }
+        private byte? DisambiguateMiddleValue(TrieNode node, int level, int m)
+        {
+            if (node.IsLeaf)
+                return DisambiguateByteValue(node.ChildIndices[m], level);
+            else
+                return _childTrieNodes[node.ChildIndices[m]].ByteValue;
+        }
+        // This binary search implementation returns a tuple:
+        // * a bool which represents whether the search completed successfully
+        // * an int representing either:
+        //      - the index of the item searched for
+        //      - the index of the location where the item should be placed to maintain a sorted list
+        // Because of the second function, a tuple is necessary since we cannot return -1 to represent a failed search 
+        private Tuple<bool, int> BinarySearch(TrieNode node, int level, byte? value)
```

Comment:
Tuple is a class. It causes an allocation. You should return a custom struct with two fields.